### PR TITLE
UndoManager: port to Vala

### PR DESCRIPTION
**Pull request by tintou**
It makes it a little bit more readable and memory-safe

### CODE REVIEW

File: libcore/UndoManager.vala [hunk 1]
```diff
@@ -0,0 +1,919 @@
...
+
+        is_valid = true;
+    }
+
+    private unowned string get_first_target_short_name () {
+        return destinations.first ().data;
+    }
+
+    private string get_uri_parent_path (string uri) {
+        return GLib.File.new_for_uri (uri).get_parent ().get_path ();
```
Comment (jeremypw) [inline libcore/UndoManager.vala]: .get_parent () could theoretically return null.

Comment (tintou) [reply libcore/UndoManager.vala]: Indeed, it wasn't taken into account in the original code but I added a workaround for this issue

**Comment by jeremypw** [comment]
@tintou Have you made any progress regarding the above crash?  I looked at it a little while - it seems to be a ref counting issue of some kind but I did not track it down.

**Comment by tintou** [comment]
@jeremypw I just pushed a fix for this

**Comment by jeremypw** [comment]
@tintou :  I am afraid it crashes at a different place now: 
```
** (io.elementary.files:6218): CRITICAL **: 17:01:19.130: marlin_undo_manager_data_add_origin_target_pair: assertion 'self != NULL' failed

Thread 1 "io.elementary.f" received signal SIGSEGV, Segmentation fault.
0x00007ffff7956a70 in marlin_undo_manager_undo_redo_done_transfer_callback (self=0x100000007, action=0x1)
    at /media/jeremy/Shared/shared_data/Vala/Elementary/pantheon-files-branches/files/libcore/UndoManager.vala:846
846	        if (!action.freed) {
(gdb) bt
#0  0x00007ffff7956a70 in marlin_undo_manager_undo_redo_done_transfer_callback (self=0x100000007, action=0x1)
    at /media/jeremy/Shared/shared_data/Vala/Elementary/pantheon-files-branches/files/libcore/UndoManager.vala:846
#1  0x00007ffff79544b5 in ____lambda36_ (_data15_=0x555555e6c120) at /media/jeremy/Shared/shared_data/Vala/Elementary/pantheon-files-branches/files/libcore/UndoManager.vala:631
#2  0x00007ffff79544d0 in _____lambda36__gcallback (self=0x555555e6c120)
    at /media/jeremy/Shared/shared_data/Vala/Elementary/pantheon-files-branches/files/libcore/UndoManager.vala:631
#3  0x00007ffff796bff1 in move_job_done (user_data=0x555555f638f0) at ../libcore/marlin-file-operations.c:4989
#4  0x00007ffff70452cf in  () at /usr/lib/x86_64-linux-gnu/libgio-2.0.so.0
#5  0x00007ffff761e185 in g_main_context_dispatch () at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#6  0x00007ffff761e550 in  () at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#7  0x00007ffff761e5dc in g_main_context_iteration () at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#8  0x00007ffff7081ddd in g_application_run () at /usr/lib/x86_64-linux-gnu/libgio-2.0.so.0
#9  0x000055555557b1b3 in _vala_main (args=0x7fffffffdf88, args_length1=1) at /media/jeremy/Shared/shared_data/Vala/Elementary/pantheon-files-branches/files/src/main.vala:26
#10 0x000055555557b1f6 in main (argc=1, argv=0x7fffffffdf88) at /media/jeremy/Shared/shared_data/Vala/Elementary/pantheon-files-branches/files/src/main.vala:16

```

I moved an empty file from one directory to another then pressed `<Ctrl>Z`.

**Comment by jeremypw** [comment]
@tintou: Did you confirm the crash I found?

**Comment by jeremypw** [comment]
Crash only occurs for move, not for copy or link.

**Comment by jeremypw** [comment]
A crash is still occurring on undoing a move:
```
** (io.elementary.files:9175): CRITICAL **: 18:08:36.105: marlin_undo_manager_data_add_origin_target_pair: assertion 'self != NULL' failed

Thread 1 "io.elementary.f" received signal SIGSEGV, Segmentation fault.
0x00007ffff7955a06 in marlin_undo_manager_undo_redo_done_transfer_callback (self=0x100000007, action=0x1) at /media/jeremy/Shared/shared_data/Vala/Elementary/pantheon-files-branches/files/libcore/UndoManager.vala:846
846	        if (!action.freed) {
(gdb) bt
#0  0x00007ffff7955a06 in marlin_undo_manager_undo_redo_done_transfer_callback (self=0x100000007, action=0x1) at /media/jeremy/Shared/shared_data/Vala/Elementary/pantheon-files-branches/files/libcore/UndoManager.vala:846
#1  0x00007ffff795344b in ____lambda36_ (_data15_=0x555556011920) at /media/jeremy/Shared/shared_data/Vala/Elementary/pantheon-files-branches/files/libcore/UndoManager.vala:631
#2  0x00007ffff7953466 in _____lambda36__gcallback (self=0x555556011920) at /media/jeremy/Shared/shared_data/Vala/Elementary/pantheon-files-branches/files/libcore/UndoManager.vala:631
#3  0x00007ffff796b94d in move_job_done (user_data=0x555555a40050) at ../libcore/marlin-file-operations.c:4989
#4  0x00007ffff70422ef in  () at /usr/lib/x86_64-linux-gnu/libgio-2.0.so.0
#5  0x00007ffff761b285 in g_main_context_dispatch () at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#6  0x00007ffff761b650 in  () at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#7  0x00007ffff761b6dc in g_main_context_iteration () at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#8  0x00007ffff707ee3d in g_application_run () at /usr/lib/x86_64-linux-gnu/libgio-2.0.so.0
#9  0x000055555557b5cb in _vala_main (args=0x7fffffffdee8, args_length1=1) at /media/jeremy/Shared/shared_data/Vala/Elementary/pantheon-files-branches/files/src/main.vala:26
#10 0x000055555557b60e in main (argc=1, argv=0x7fffffffdee8) at /media/jeremy/Shared/shared_data/Vala/Elementary/pantheon-files-branches/files/src/main.vala:16
```

**Comment by jeremypw** [comment]
It may be simpler to rewrite Marlin.FileOperations in Vala as well :-(

**Comment by danrabbit** [comment]
Closing since it seems that https://github.com/elementary/files/pull/1132 supersedes this

**Pull request closed without merging**